sdk/rust/oak/src/lib.rs

```diff
@@ -149,34 +176,69 @@ pub fn channel_read(half: ReadHandle, buf: &mut Vec<u8>, handles: &mut Vec<Handl
                 buf.as_mut_ptr(),
                 buf.capacity(),
                 &mut actual_size,
-                handles.as_mut_ptr() as *mut u8,
-                handles.capacity() as u32,
+                handles_buf.as_mut_ptr(),
+                handles_buf.capacity() as u32 / 8, // Handle count, not byte count
                 &mut actual_handle_count,
             ) as i32
         });
+
         match status {
             Some(s) => match s {
                 OakStatus::OK | OakStatus::ERR_CHANNEL_EMPTY => {
                     unsafe {
                         buf.set_len(actual_size as usize);
-                        // Handles are written in little-endian order, which matches Wasm spec.
-                        handles.set_len(actual_handle_count as usize);
+
+                        // actual_handle_count is number of handles not bytes
+                        handles_buf.set_len(actual_handle_count as usize * 8);
+                        Handle::unpack(&handles_buf, actual_handle_count, handles);
                     };
                     return s;
                 }
-                OakStatus::ERR_BUFFER_TOO_SMALL if !(*resized) => {
+
+                // ERR_BUFFER_TOO_SMALL can shadow ERR_HANDLE_SPACE_TOO_SMALL
+                OakStatus::ERR_BUFFER_TOO_SMALL | OakStatus::ERR_HANDLE_SPACE_TOO_SMALL
+                    if !(*resized) =>
+                {
```

Comment:
We should sync up on whether to pull in #453 first or just use this &ndash; thoughts?

Reply:
Rebasing on  #453